Lib/Common/Core/Util/AsyncCounterEvent.cs

```diff
@@ -0,0 +1,72 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="AsyncCounterEvent.cs" company="Microsoft">
+//    Copyright 2017 Microsoft Corporation
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+// </copyright>
+//-----------------------------------------------------------------------
+
+namespace Microsoft.WindowsAzure.Storage.Core.Util
+{
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    internal sealed class AsyncCounterEvent
+    {
+        private TaskCompletionSource<bool> taskCompletionSource = new TaskCompletionSource<bool>();
```

Comment:
Would it be "safer" to use `TaskContinuationOptions.RunAsynchronously` here?

Reply:
Yes, we definitely need this, although I think it's 'TaskCreationOptions', not 'TaskContinuationOptions'.  Looks like this will require an upgrade to .NET 4.6 as well.

Reply:
We are waiting on usage data before we can make a final decision about upgrading to 4.6 for NetFx target. But even if that doesn't happen we can manage to keep this change for the RT/NetCore codebase and use the existing version for Desktop.

Reply:
Correct it would require a framework upgrade